core/src/smartcontracts/isi/asset.rs

```diff
@@ -54,89 +54,165 @@ pub mod isi {
         }
     }
 
-    impl<W: WorldTrait> Execute<W> for Mint<Asset, u32> {
-        type Error = Error;
+    /// Trait for asset values that could be converted to `f64` and observed by prometheus.
+    trait FloatMetricFrom {
+        fn from(value: Self) -> f64;
+    }
 
-        #[metrics(+"mint_qty")]
-        fn execute(
-            self,
-            _authority: <Account as Identifiable>::Id,
-            wsv: &WorldStateView<W>,
-        ) -> Result<(), Self::Error> {
-            let asset_id = self.destination_id;
+    impl FloatMetricFrom for u128 {
+        #[allow(clippy::cast_precision_loss)]
+        fn from(value: Self) -> f64 {
+            value as f64
+        }
+    }
 
-            assert_can_mint(&asset_id.definition_id, wsv, AssetValueType::Quantity)?;
-            wsv.asset_or_insert(&asset_id, 0_u32)?;
-            wsv.modify_asset(&asset_id, |asset| {
-                let quantity: &mut u32 = asset
-                    .try_as_mut()
-                    .map_err(eyre::Error::from)
-                    .map_err(|e| Error::Conversion(e.to_string()))?;
-                *quantity = quantity
-                    .checked_add(self.object)
-                    .ok_or(MathError::Overflow)?;
-                wsv.metrics.tx_amounts.observe(f64::from(*quantity));
+    impl FloatMetricFrom for u32 {
+        fn from(value: Self) -> f64 {
+            value.into()
+        }
+    }
 
-                Ok(AssetEvent::Added(asset_id.clone()))
-            })
+    impl FloatMetricFrom for Fixed {
+        fn from(value: Self) -> f64 {
+            value.into()
         }
     }
 
-    impl<W: WorldTrait> Execute<W> for Mint<Asset, u128> {
-        type Error = Error;
+    /// Trait for `try_checked_add` operation on an asset value that can fail and return [`Error`].
+    trait TryCheckedAdd {
+        fn try_checked_add(self, rhs: Self) -> Result<Self, Error>
+        where
+            Self: Sized;
+    }
 
-        #[metrics(+"mint_big_qty")]
-        fn execute(
-            self,
-            _authority: <Account as Identifiable>::Id,
-            wsv: &WorldStateView<W>,
-        ) -> Result<(), Self::Error> {
-            let asset_id = self.destination_id;
+    impl TryCheckedAdd for u32 {
```

Comment:
These traits could be moved to `iroha_data_model` or `iroha_data_primitives`. That would be more in line with our upcoming refactor of the data model.